File: backend/channel_manager.py
"""
Channel manager: reads and writes Telegram/Slack channel config on a running
Daytona sandbox via the OpenClaw CLI.

All functions are synchronous (blocking) — call them from a thread pool
executor in async FastAPI route handlers.
"""

import json
import time

from daytona import Daytona, DaytonaConfig

DAYTONA_API_URL = "https://app.daytona.io/api"
GATEWAY_PORT = 18789
CONFIG_PATH = "/root/.openclaw/openclaw.json"


# ── Sandbox helpers ────────────────────────────────────────────────────────────

def _get_sandbox(api_key: str, sandbox_id: str):
    cfg = DaytonaConfig(api_key=api_key, api_url=DAYTONA_API_URL)
    return Daytona(cfg).get(sandbox_id)


def _exec(sandbox, cmd: str, timeout: int = 30) -> tuple[bool, str]:
    res = sandbox.process.exec(cmd, timeout=timeout)
    return res.exit_code == 0, (res.result or "").strip()


def _read_openclaw_config(sandbox) -> dict:
    ok, out = _exec(
        sandbox,
        f"node -e \"process.stdout.write(require('fs').readFileSync('{CONFIG_PATH}','utf8'))\"",
    )
    if not ok or not out:
        return {}
    try:
        return json.loads(out)
    except Exception:
        return {}


def _set_cfg(sandbox, dotted_key: str, value) -> tuple[bool, str]:
    """Run `openclaw config set <key> <value>` safely."""
    if isinstance(value, bool):
        val_str = "true" if value else "false"
        return _exec(sandbox, f"openclaw config set {dotted_key} {val_str}")
    else:
        # Escape single quotes in value, wrap in single quotes for the shell
        safe = str(value).replace("'", "'\\''")
        return _exec(sandbox, f"openclaw config set {dotted_key} '{safe}'")


def _restart_gateway(sandbox) -> None:
    _exec(sandbox, "openclaw gateway stop 2>/dev/null || true", timeout=15)
    time.sleep(2)
    _exec(
        sandbox,
        f"nohup openclaw gateway run --bind lan --port {GATEWAY_PORT} "
        f"> /tmp/openclaw-gateway.log 2>&1 &",
        timeout=10,
    )


def _mask(v: str) -> str:
    """Return a masked version of a credential string."""
    if not v:
        return ""
    if len(v) <= 8:
        return "***"
    return v[:4] + "***" + v[-4:]


# ── Public API ─────────────────────────────────────────────────────────────────

def get_channels_config(api_key: str, sandbox_id: str) -> dict:
    """
    Read channel config from the sandbox, returning masked credential values.
    """
    sb = _get_sandbox(api_key, sandbox_id)
    config = _read_openclaw_config(sb)
    channels = config.get("channels", {})

    tg = channels.get("telegram", {})
    sl = channels.get("slack", {})

    return {
        "telegram": {
            "enabled": bool(tg.get("enabled", False)),
            "botToken": _mask(tg.get("botToken", "")),
            "dmPolicy": tg.get("dmPolicy", "pairing"),
        },
        "slack": {
            "enabled": bool(sl.get("enabled", False)),
            "mode": sl.get("mode", "socket"),
            "appToken": _mask(sl.get("appToken", "")),
            "botToken": _mask(sl.get("botToken", "")),
            "signingSecret": _mask(sl.get("signingSecret", "")),
            "dmPolicy": sl.get("dmPolicy", "pairing"),
        },
    }


def set_telegram_config(api_key: str, sandbox_id: str, cfg: dict) -> dict:
    """
    Apply Telegram channel settings on the sandbox and restart the gateway.

    Fields in `cfg`:
      enabled   bool
      botToken  str  (skip update if empty/None — preserves existing token)
      dmPolicy  str  pairing | allowlist | open | disabled
    """
    sb = _get_sandbox(api_key, sandbox_id)
    logs: list[str] = []

    if "enabled" in cfg:
        ok, _ = _set_cfg(sb, "channels.telegram.enabled", cfg["enabled"])
        logs.append(f"{'✓' if ok else '✗'} enabled={cfg['enabled']}")

    if cfg.get("botToken"):
        ok, _ = _set_cfg(sb, "channels.telegram.botToken", cfg["botToken"])
        logs.append(f"{'✓' if ok else '✗'} botToken=***")

    if cfg.get("dmPolicy"):
        ok, _ = _set_cfg(sb, "channels.telegram.dmPolicy", cfg["dmPolicy"])
        logs.append(f"{'✓' if ok else '✗'} dmPolicy={cfg['dmPolicy']}")

    _restart_gateway(sb)
    logs.append("✓ Gateway restarted")

    return {"ok": True, "logs": logs}


def set_slack_config(api_key: str, sandbox_id: str, cfg: dict) -> dict:
    """
    Apply Slack channel settings on the sandbox and restart the gateway.

    Fields in `cfg`:
      enabled       bool
      mode          str   socket | http
      appToken      str   xapp-...  (skip if empty)
      botToken      str   xoxb-...  (skip if empty)
      signingSecret str             (skip if empty)
      dmPolicy      str   pairing | allowlist | open | disabled
    """
    sb = _get_sandbox(api_key, sandbox_id)
    logs: list[str] = []

    if "enabled" in cfg:
        ok, _ = _set_cfg(sb, "channels.slack.enabled", cfg["enabled"])
        logs.append(f"{'✓' if ok else '✗'} enabled={cfg['enabled']}")

    if cfg.get("mode"):
        ok, _ = _set_cfg(sb, "channels.slack.mode", cfg["mode"])
        logs.append(f"{'✓' if ok else '✗'} mode={cfg['mode']}")

    if cfg.get("appToken"):
        ok, _ = _set_cfg(sb, "channels.slack.appToken", cfg["appToken"])
        logs.append(f"{'✓' if ok else '✗'} appToken=***")

    if cfg.get("botToken"):
        ok, _ = _set_cfg(sb, "channels.slack.botToken", cfg["botToken"])
        logs.append(f"{'✓' if ok else '✗'} botToken=***")

    if cfg.get("signingSecret"):
        ok, _ = _set_cfg(sb, "channels.slack.signingSecret", cfg["signingSecret"])
        logs.append(f"{'✓' if ok else '✗'} signingSecret=***")

    if cfg.get("dmPolicy"):
        ok, _ = _set_cfg(sb, "channels.slack.dmPolicy", cfg["dmPolicy"])
        logs.append(f"{'✓' if ok else '✗'} dmPolicy={cfg['dmPolicy']}")

    _restart_gateway(sb)
    logs.append("✓ Gateway restarted")

    return {"ok": True, "logs": logs}


def probe_channel_status(api_key: str, sandbox_id: str, channel: str) -> dict:
    """
    Run `openclaw channels status --probe` and return parsed output.
    `channel` is used only to label the result; the command probes all channels.
    """
    sb = _get_sandbox(api_key, sandbox_id)
    ok, output = _exec(sb, "openclaw channels status --probe 2>&1", timeout=45)
    return {"ok": ok, "channel": channel, "output": output}


# ── Pairing ────────────────────────────────────────────────────────────────────

def list_pairing_requests(api_key: str, sandbox_id: str, channel: str) -> dict:
    """
    Run `openclaw pairing list <channel>` and return the raw output plus any
    codes parsed from lines that look like `XXXX1234`.
    """
    sb = _get_sandbox(api_key, sandbox_id)
    ok, output = _exec(sb, f"openclaw pairing list {channel} 2>&1", timeout=30)

    # Try to extract 8-char uppercase alphanumeric codes from the output
    import re
    codes = re.findall(r"\b([A-Z0-9]{8})\b", output)

    return {"ok": ok, "channel": channel, "output": output, "codes": codes}


def approve_pairing(api_key: str, sandbox_id: str, channel: str, code: str) -> dict:
    """
    Run `openclaw pairing approve <channel> <code>` and return success/failure.
    """
    # Sanitise code — only allow uppercase alphanumeric to prevent injection
    import re
    clean_code = re.sub(r"[^A-Z0-9]", "", code.upper())
    if not clean_code:
        return {"ok": False, "output": "Invalid pairing code"}

    sb = _get_sandbox(api_key, sandbox_id)
    ok, output = _exec(sb, f"openclaw pairing approve {channel} {clean_code} 2>&1", timeout=30)
    return {"ok": ok, "channel": channel, "code": clean_code, "output": output}

File: ruh-backend/src/channelManager.ts
/**
 * Channel manager: reads and writes Telegram/Slack channel config on a running
 * Docker sandbox container via the OpenClaw CLI.
 */

import { dockerExec, getContainerName } from './sandboxManager';

const GATEWAY_PORT = 18789;
const CONFIG_PATH = '/root/.openclaw/openclaw.json';

// ── Container helpers ──────────────────────────────────────────────────────────

async function execCmd(
  sandboxId: string,
  cmd: string,
  timeoutSec = 30,
): Promise<[boolean, string]> {
  return dockerExec(getContainerName(sandboxId), cmd, timeoutSec * 1000);
}

async function readOpenclawConfig(sandboxId: string): Promise<Record<string, unknown>> {
  const [ok, out] = await execCmd(
    sandboxId,
    `node -e "process.stdout.write(require('fs').readFileSync('${CONFIG_PATH}','utf8'))"`,
  );
  if (!ok || !out) return {};
  try {
    return JSON.parse(out) as Record<string, unknown>;
  } catch {
    return {};
  }
}

async function setCfg(
  sandboxId: string,
  dottedKey: string,
  value: unknown,
): Promise<[boolean, string]> {
  if (typeof value === 'boolean') {
    return execCmd(sandboxId, `openclaw config set ${dottedKey} ${value ? 'true' : 'false'}`);
  }
  const safe = String(value).replace(/'/g, "'\\''");
  return execCmd(sandboxId, `openclaw config set ${dottedKey} '${safe}'`);
}

async function restartGateway(sandboxId: string): Promise<void> {
  await execCmd(sandboxId, 'openclaw gateway stop 2>/dev/null || true', 15);
  await Bun.sleep(2000);
  await execCmd(
    sandboxId,
    `nohup openclaw gateway run --bind lan --port ${GATEWAY_PORT} > /tmp/openclaw-gateway.log 2>&1 &`,
    10,
  );
}

function mask(v: string): string {
  if (!v) return '';
  if (v.length <= 8) return '***';
  return v.slice(0, 4) + '***' + v.slice(-4);
}

// ── Public API ─────────────────────────────────────────────────────────────────

export async function getChannelsConfig(
  sandboxId: string,
): Promise<Record<string, unknown>> {
  const config = await readOpenclawConfig(sandboxId);
  const channels = (config['channels'] ?? {}) as Record<string, Record<string, unknown>>;

  const tg = channels['telegram'] ?? {};
  const sl = channels['slack'] ?? {};

  return {
    telegram: {
      enabled: Boolean(tg['enabled'] ?? false),
      botToken: mask(String(tg['botToken'] ?? '')),
      dmPolicy: tg['dmPolicy'] ?? 'pairing',
    },
    slack: {
      enabled: Boolean(sl['enabled'] ?? false),
      mode: sl['mode'] ?? 'socket',
      appToken: mask(String(sl['appToken'] ?? '')),
      botToken: mask(String(sl['botToken'] ?? '')),
      signingSecret: mask(String(sl['signingSecret'] ?? '')),
      dmPolicy: sl['dmPolicy'] ?? 'pairing',
    },
  };
}

export async function setTelegramConfig(
  sandboxId: string,
  cfg: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  const logs: string[] = [];

  if ('enabled' in cfg) {
    const [ok] = await setCfg(sandboxId, 'channels.telegram.enabled', cfg['enabled']);
    logs.push(`${ok ? '✓' : '✗'} enabled=${cfg['enabled']}`);
  }
  if (cfg['botToken']) {
    const [ok] = await setCfg(sandboxId, 'channels.telegram.botToken', cfg['botToken']);
    logs.push(`${ok ? '✓' : '✗'} botToken=***`);
  }
  if (cfg['dmPolicy']) {
    const [ok] = await setCfg(sandboxId, 'channels.telegram.dmPolicy', cfg['dmPolicy']);
    logs.push(`${ok ? '✓' : '✗'} dmPolicy=${cfg['dmPolicy']}`);
  }

  await restartGateway(sandboxId);
  logs.push('✓ Gateway restarted');
  return { ok: true, logs };
}

export async function setSlackConfig(
  sandboxId: string,
  cfg: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  const logs: string[] = [];

  if ('enabled' in cfg) {
    const [ok] = await setCfg(sandboxId, 'channels.slack.enabled', cfg['enabled']);
    logs.push(`${ok ? '✓' : '✗'} enabled=${cfg['enabled']}`);
  }
  if (cfg['mode']) {
    const [ok] = await setCfg(sandboxId, 'channels.slack.mode', cfg['mode']);
    logs.push(`${ok ? '✓' : '✗'} mode=${cfg['mode']}`);
  }
  if (cfg['appToken']) {
    const [ok] = await setCfg(sandboxId, 'channels.slack.appToken', cfg['appToken']);
    logs.push(`${ok ? '✓' : '✗'} appToken=***`);
  }
  if (cfg['botToken']) {
    const [ok] = await setCfg(sandboxId, 'channels.slack.botToken', cfg['botToken']);
    logs.push(`${ok ? '✓' : '✗'} botToken=***`);
  }
  if (cfg['signingSecret']) {
    const [ok] = await setCfg(sandboxId, 'channels.slack.signingSecret', cfg['signingSecret']);
    logs.push(`${ok ? '✓' : '✗'} signingSecret=***`);
  }
  if (cfg['dmPolicy']) {
    const [ok] = await setCfg(sandboxId, 'channels.slack.dmPolicy', cfg['dmPolicy']);
    logs.push(`${ok ? '✓' : '✗'} dmPolicy=${cfg['dmPolicy']}`);
  }

  await restartGateway(sandboxId);
  logs.push('✓ Gateway restarted');
  return { ok: true, logs };
}

export async function probeChannelStatus(
  sandboxId: string,
  channel: string,
): Promise<Record<string, unknown>> {
  const [ok, output] = await execCmd(sandboxId, 'openclaw channels status --probe 2>&1', 45);
  return { ok, channel, output };
}

// ── Pairing ────────────────────────────────────────────────────────────────────

export async function listPairingRequests(
  sandboxId: string,
  channel: string,
): Promise<Record<string, unknown>> {
  const [ok, output] = await execCmd(sandboxId, `openclaw pairing list ${channel} 2>&1`, 30);
  const codes = [...output.matchAll(/\b([A-Z0-9]{8})\b/g)].map((m) => m[1]);
  return { ok, channel, output, codes };
}

export async function approvePairing(
  sandboxId: string,
  channel: string,
  code: string,
): Promise<Record<string, unknown>> {
  const cleanCode = code.toUpperCase().replace(/[^A-Z0-9]/g, '');
  if (!cleanCode) return { ok: false, output: 'Invalid pairing code' };

  const [ok, output] = await execCmd(
    sandboxId,
    `openclaw pairing approve ${channel} ${cleanCode} 2>&1`,
    30,
  );
  return { ok, channel, code: cleanCode, output };
}
